notes/advanced-data-types.md

```diff
@@ -0,0 +1,530 @@
+    // Error: Good has already been defined by People
+    
+    int Meh = 0;
+    // Error: Bad is already a constant
+} // End Others Scope
+
+enum Professors = {Okay, Meh} // Begin Professors Scope
+// Okay: the enum Others is out of scope of Professors
+
+// End People Scope
+// End Professors Scope
+```
+<div style="text-align: center"> Figure 5a - Enum Errors </div>
+
+Another bad thing is attempting to compare different enums when you don't want it. This is because C++ is not type-safe, so it automatically convert the `enum` constants to integers in a process known as *implicit conversion*. The following program functions, but doesn't conceptually make any sense:
+
+```
+weekday day = Friday;
+painMeter painRating = pain;
+if(painRating == painRating) {
+    std::cout << "Days are pain!!!" << std::endl;
+}
+
+// Output: Days are pain!!!
+```
+<div style="text-align: center"> Figure 5b - Demonstration of Lack of Type-Safety for Enums </div>
+
+You see, one should not be able to compare these two enums since they are unrelated, but since friday and pain are both 5, they are "equal". To avoid this, C++11 introduces the `enum class`, which is a strongly typed and strongly scoped type.
+T
+```
+/*
+enum class enumName {
+    constant1,
+    constant2,
+    ...
+    constantN
+};
+*/
+
+enum class betterWeekday {
+    betterSunday,
+    betterMonday,
+    betterTuesday,
+    betterWednesday,
+    betterThursday,
+    betterFriday,
+    betterSaturday
+};
+```
+<div style="text-align: center"> Figure 6a - Enum Class Declaration </div>
+
+To reference a constant within an `enum class`, you must use the scope resolution operator (`::`):
+
+<div style="text-align: center"> 
+
+`enumClassName::Field` 
+
+</div>
+
+This tells us that constants belonging to `enum class`es are specific to those classes, meaning that we can now define variables or other `enum class` constants with the same name. To demonstrate, let's draw an analogy to Figure 5a with the following valid example:
+
+```
+enum class People = {Good, Bad};
+
+int main() {
+    enum class Others = {Good, Meh}; // Okay now, since People::Good is not Others::Good
+    
+    int Meh = 0; // This is okay now, since main::Meh is not Others::Meh
+}
+
+enum Professors = {Okay, Meh} // Still okay
+```
+<div style="text-align: center"> Figure 7a - Demonstration of the Strongly Scoped Enum Class </div>
+
+However, this means the initialization of the `enum class` is different
+
+```
+// enumClassName enumClassVarName = enumClassName::Constant;
+
+betterWeekday betterDay = betterWeekday::betterWednesday;
+
+```
+<div style="text-align: center"> Figure 6b - Enum Class Initialization </div>
+
+With the use of the keyword `class`, C++ establishes `enum class`es as distinct entities, which does not allow for automatic type conversion.
+
+```
+// Pretend weekday and painMeter are now enum classes
+weekday day = weekday::Friday;
+painMeter painRating = painMeter::pain;
+if(painRating == painRating) {
```

Comment:
Again, you don't seem to be using `day`, just comparing `painRating` with itself, which should always be true